spec/light-client/supervisor/supervisor_001_draft.md

```diff
@@ -463,69 +463,93 @@ func Sequential-Supervisor (initdata LCInitData) (Error) {
+increase trust since proving a conflicting block is a 
+[light client attack](https://informal.systems) 
+and not just a [bogus](https://informal.systems) block could result in
+performing backwards verification beyond the trusted period, thus a fruitless
+endeavour.
+
+However, with the notion of it's better to fail earlier than later, the golang
+light client implementation will perform a consistency check on all providers
+and will error if one returns a different header, allowing the user
+the opportunity to reinitialize.
 
 #### **[LC-FUNC-INIT.1]:**
 
 ```go
-func InitLightClient (initData LCInitData) (LightStore, Error) {
+func InitLightClient(initData LCInitData) (LightStore, Error) {
+    var initialBlock LightBlock
 
-    if LCInitData.LightBlock != nil {
-        // we trust the provided initial block.
-        newblock := LCInitData.LightBlock
-    }
-    else {
-        genesisBlock := makeblock(initData.genesisDoc);
+    switch {
+    case LCInitData.TrustedBlock != nil:
+        // we trust the block from a prior initialization
+        initialBlock = LCInitData.TrustedBlock
 
-        result := NoResult;
-        while result != ResultSuccess {
-            current = FetchLightBlock(PeerList.primary(), genesisBlock.Header.Height + 1)
-            // QUESTION: is the height with "+1" OK?
+    case LCInitData.TrustedHash != nil:
+        untrustedBlock := FetchLightBlock(PeerList.Primary(), LCInitData.TrustedHeight)
+        
 
-            if CANNOT_VERIFY = ValidAndVerify(genesisBlock, current) {
-                Replace_Primary();
-            }
-            else {
-                result = ResultSuccess
-            }
-        }
-  
-        // cross-check
-  auxLS := new LightStore
-  auxLS.Add(current)
-        Evidences := AttackDetector(genesisBlock, auxLS)
-        if Evidences.Empty {
-            newBlock := current
+        // verify that the hashes match
+        if untrustedBlock.Hash() != LCInitData.TrustedHash {
+            return nil, Error("Primary returned block with different hash")
         }
-        else {
-            // [LC-SUMBIT-EVIDENCE.1]
-            submitEvidence(Evidences);
-            return(nil, ErrorAttack);
+        // after checking the hash we now trust the block
+        initialBlock = untrustedBlock        
+    }
+    case LCInitData.Genesis != nil:
+        untrustedBlock := FetchLightBlock(PeerList.Primary(), LCInitData.Genesis.InitialHeight)
+        
+        // verify that 2/3+ of the validator set signed the untrustedBlock
+        if err := VerifyCommit(untrustedBlock.Commit, LCInitData.Genesis.Validators); err != nil {
+            return nil, err
         }
+
+        // we can now trust the block
+        initialBlock = untrustedBlock
+    default:
+        return nil, Error("No initial data was provided")
+
+    if err := CrossCheck(initialBlock, PeerList.Witnesses()); err != nil {
+        return nil, err
     }
```

Comment:
I could remove this cross check from the spec pseudocode as it's not really important to the correctness of the spec. I think `[LC-FUNC-INIT.1]` would need to be updated to  `[LC-FUNC-INIT.2]` anyway because there are still changes to the logic

Reply:
OK. Perhaps we could also add a comment within the pseudo code that that line is optional and done in the golang version. I am a bit under water right now and will do a closer check of the new version in the week of July 26